lib/profileAgents.js

```diff
@@ -0,0 +1,361 @@
+const KMS_MODULE = 'ssm-v1';
+
+bedrock.events.on('bedrock-mongodb.ready', async () => {
+  await promisify(database.openCollections)(['profile-profileAgent']);
+
+  await promisify(database.createIndexes)([{
+    collection: 'profile-profileAgent',
+    fields: {'profileAgent.id': 1},
+    options: {unique: true, background: false}
+  }, {
+    collection: 'profile-profileAgent',
+    fields: {'profileAgent.profile': 1},
+    options: {unique: false, background: false}
+  }, {
+    collection: 'profile-profileAgent',
+    fields: {'profileAgent.controller': 1},
+    options: {unique: true, background: false}
+  }, {
+    collection: 'profile-profileAgent',
+    fields: {'profileAgent.id': 1, 'profileAgent.sequence': 1},
+    options: {unique: false, background: false}
+  }, {
+    collection: 'profile-profileAgent',
+    fields: {'profileAgent.account': 1, 'profileAgent.id': 1},
+    options: {
+      partialFilterExpression: {account: {$exists: true}},
+      unique: true,
+      background: false
+    }
+  }]);
+});
+
+/**
+ * Creates a ProfileAgent and associates it with an account if provided.
+ *
+ * @param {String} accountId the id of the account to associate with the
+ *                           ProfileAgent.
+ *
+ * @return {Promise<ProfileAgent>} resolves to a ProfileAgent.
+ */
+exports.create = async ({accountId} = {}) => {
+  // 1. Generate a random secret
+  const secret = base64url.encode(crypto.randomBytes(32));
+  const handle = 'primary';
+  // 2. Generate capability agent for the zCap key
+  const capabilityAgent = await CapabilityAgent.fromSecret({handle, secret});
+  // 3. Create keystore in order to create the zCap key for the profile agent
+  // TODO: Will want to store the capability agent up here first, otherwise we
+  //       have no clean way of cleaning up the other stuff that gets created
+  //       on other systems should we experience a failure.
+  const keystore = await kms.createKeystore(
+    {capabilityAgent, referenceId: 'primary'});
+  // 4. Create the zCap key for the profile agent
+  const keystoreAgent = kms.getKeystoreAgent({capabilityAgent, keystore});
+  const key = await keystoreAgent.generateKey(
+    {type: 'Ed25519VerificationKey2018', kmsModule: KMS_MODULE});
+  // 5. Use the zCap key to create the DID for the profile agent
+  await utils.setKeyId({key});
+  // 6. Generate capability agent for ProfileAgent
+  const paZcapAgent = new CapabilityAgent({handle: 'primary', signer: key});
+  // 7. Create keystore to store keys controlled by the profile agent
+  const paKeystore = await kms.createKeystore(
+    {capabilityAgent: paZcapAgent, referenceId: 'primary'});
+  const paKeystoreAgent = kms.getKeystoreAgent(
+    {capabilityAgent: paZcapAgent, keystore: paKeystore});
+
+  const did = key.id.split('#')[0];
+  const profileAgent = {
+    id: did,
+    sequence: 0,
+    account: accountId,
+    controller: {
+      id: capabilityAgent.id,
+      seed: secret,
+      keystore: keystore.id
```

Comment:
I'm not sure how I feel about this embedding of controller information like this, but maybe it's good. Just making a note.